etrago/cluster/gasclustering.py

```diff
@@ -0,0 +1,397 @@
+        custom_strategies=bus_strategies,
+    )
+    new_buses.index.name = "bus_id"
+
+    io.import_components_from_dataframe(network_gasgrid_c, new_buses, "Bus")
+
+    if with_time:
+        network_gasgrid_c.snapshot_weightings = network.snapshot_weightings.copy()
+        network_gasgrid_c.set_snapshots(network.snapshots)
+
+    # Aggregate one port components
+    one_port_components = ["Generator", "Load", "Store"]
+
+    for one_port in one_port_components:
+        one_port_components.remove(one_port)
+        new_df, new_pnl = aggregateoneport(
+            network,
+            busmap,
+            component=one_port,
+            with_time=with_time,
+            custom_strategies=one_port_strategies.get(one_port, {}),
+        )
+        io.import_components_from_dataframe(network_gasgrid_c, new_df, one_port)
+        for attr, df in iteritems(new_pnl):
+            io.import_series_from_dataframe(network_gasgrid_c, df, one_port, attr)
+
+    for c in network.iterate_components(one_port_components):
+        io.import_components_from_dataframe(
+            network_gasgrid_c,
+            c.df.assign(bus=c.df.bus.map(busmap)).dropna(subset=["bus"]),
+            c.name,
+        )
+
+    if with_time:
+        for c in network.iterate_components(one_port_components):
+            for attr, df in iteritems(c.pnl):
+                if not df.empty:
+                    io.import_series_from_dataframe(network_gasgrid_c, df, c.name, attr)
+
+    # Aggregate links
+    new_links = (
+        network.links.assign(
+            bus0=network.links.bus0.map(busmap), bus1=network.links.bus1.map(busmap)
+        )
+        .dropna(subset=["bus0", "bus1"])
+        .loc[lambda df: df.bus0 != df.bus1]
+    )
+
+    new_links["link_id"] = new_links.index
+
+    strategies = {
+        "p_nom": "sum",
+        "length": "mean",
+    }
+    strategies.update(
+        {col: "first" for col in new_links.columns if col not in strategies}
+    )
+
+    gas_carriers = [  # This list should be replace by an automatic selection
+        "CH4",
+        "CH4_to_H2",
+        "H2_feedin",
+        "H2_ind_load",
+        "H2_to_CH4",
+        "H2_to_power",
+        "power_to_H2",
+        "central_gas_CHP",
+        "central_gas_CHP_heat",
+        "industrial_gas_CHP",
+        "rural_gas_boiler",
+        "central_gas_boiler",
+        "OCGT",
+    ]
+
+    gas_links = new_links[new_links["carrier"].isin(gas_carriers)].copy()
+
+    combinations = gas_links.groupby(["bus0", "bus1", "carrier"]).agg(strategies)
+    combinations.reset_index(drop=True, inplace=True)
+
+    combinations["buscombination"] = combinations[["bus0", "bus1"]].apply(
+        lambda x: tuple(sorted([str(x.bus0), str(x.bus1)])), axis=1
+    )
+
+    strategies.update(
+        {col: "first" for col in combinations.columns if col not in strategies}
+    )
+
+    combinations_final = combinations.groupby(["buscombination", "carrier"]).agg(
+        strategies
+    )

```

Comment:
I'm wondering why you are clustering the links twice (line 286 and line 297). Do you want to cluster links bidirectional?

Reply:
But there could be problems when you have unidirectional links which should be not combined with links going into the other direction. If you have in the network with the full complexity unidirectional links in both directories, you don't want to combine them and have only one link into one direction. Or am I mistaken?